Tools/AP_Periph/serial_tunnel.cpp

```diff
@@ -0,0 +1,210 @@
+    if (uart_num == -1) {
+        uart_num = g.adsb_port;
+    }
+#endif
+#if HAL_PROXIMITY_ENABLED
+    if (uart_num == -1) {
+        uart_num = g.proximity_port;
+    }
+#endif
+    return uart_num;
+}
+
+/*
+  handle tunnel data
+ */
+void AP_Periph_FW::handle_tunnel_Targetted(CanardInstance* ins, CanardRxTransfer* transfer)
+{
+    uavcan_tunnel_Targetted pkt;
+    if (uavcan_tunnel_Targetted_decode(transfer, &pkt)) {
+        return;
+    }
+    if (pkt.target_node != canardGetLocalNodeID(ins)) {
+        return;
+    }
+    if (monitor.buffer == nullptr) {
+        monitor.buffer = new ByteBuffer(1024);
+        if (monitor.buffer == nullptr) {
+            return;
+        }
+    }
+    int8_t uart_num = pkt.serial_id;
+    if (uart_num == -1) {
+        uart_num = get_default_tunnel_serial_port();
+    }
+    if (uart_num < 0) {
+        return;
+    }
+    auto *uart = hal.serial(uart_num);
+    if (uart == nullptr) {
+        return;
+    }
+    if (monitor.uart_num != uart_num && monitor.uart != nullptr) {
+        // remove monitor from previous uart
+        hal.serial(monitor.uart_num)->set_monitor_read_buffer(nullptr);
+    }
+    monitor.uart_num = uart_num;
+    if (uart != monitor.uart) {
+        // change of uart or expired, clear old data
+        monitor.buffer->clear();
+        monitor.uart = uart;
+    }
+    if (monitor.uart == nullptr) {
+        return;
+    }
+    /*
+      allow for locked state to change at any time, so users can
+      switch between locked and unlocked while connected
+     */
+    monitor.locked = (pkt.options & UAVCAN_TUNNEL_TARGETTED_OPTION_LOCK_PORT) != 0;
+    if (monitor.locked) {
+        monitor.uart->lock_port(TUNNEL_LOCK_KEY, TUNNEL_LOCK_KEY);
+    } else {
+        monitor.uart->lock_port(0,0);
+    }
+    monitor.node_id = transfer->source_node_id;
+    monitor.protocol = pkt.protocol.protocol;
+    if (pkt.baudrate != monitor.baudrate) {
+        if (monitor.locked && pkt.baudrate != 0) {
+            monitor.uart->begin_locked(pkt.baudrate, 0, 0, TUNNEL_LOCK_KEY);
+            debug("begin_locked %u", unsigned(pkt.baudrate));
+        }
+        monitor.baudrate = pkt.baudrate;
+    }
+    monitor.uart->set_monitor_read_buffer(monitor.buffer);
+    monitor.last_request_ms = AP_HAL::millis();
+
+    // write to device
+    if (pkt.buffer.len > 0) {
+        if (monitor.locked) {
+            debug("write_locked %u", unsigned(pkt.buffer.len));
+            monitor.uart->write_locked(pkt.buffer.data, pkt.buffer.len, TUNNEL_LOCK_KEY);
+        } else {
+            monitor.uart->write(pkt.buffer.data, pkt.buffer.len);
+        }
+    } else {
+        debug("locked keepalive");
+    }
+}
+
+/*
+  send tunnelled serial data
+ */
+void AP_Periph_FW::send_serial_monitor_data()
+{
+    if (monitor.uart == nullptr ||
+        monitor.node_id == 0 ||
+        monitor.buffer == nullptr) {
+        return;
+    }
+    const uint32_t last_req_ms = monitor.last_request_ms;
```

Comment:
its a good pattern to get used to which prevents a race if the timestamps updates from a callback. It does update from a callback in this case, but the callback is (for the moment) on the same thread. If/when we move AP_Periph to the model we use in AP_DroneCAN then this would matter